File: Java/2023/main/year2023/Day25/MinimumCut.java
package year2023.Day25;

import java.io.*;
import java.util.*;

import JavaDataModel.*;

@AoCSolution()
public class MinimumCut implements SolutionBase {
    void readFile(BufferedReader input) throws Exception {
        input.lines().forEach(l -> {
            var split = l.split(":");
            int inWire = Hasher(split[0].trim());
            Arrays.stream(split[1].trim().split(" ")).forEach(w -> {
                int wHash = Hasher(w);
                WireGraph.computeIfAbsent(inWire, s -> new HashSet<>()).add(wHash);
                WireGraph.computeIfAbsent(wHash, s -> new HashSet<>()).add(inWire);
            });
        });
        input.close();
    }

    int Hasher(String WireName) {
        return WireName.chars().reduce(0, (i, c) -> i * 26 + c - 'a');
    }

    Map<Integer, Set<Integer>> WireGraph = new HashMap<>();

    int CutCount(Map<Integer, Set<Integer>> Cliques) {
        if (Cliques.size() != 2)
            return -1;
        List<Integer> CliqueNames = Cliques.keySet().stream().toList();
        int cutCnt = 0;
        int from = CliqueNames.get(0), to = CliqueNames.get(1);
        for (var fromV : Cliques.get(from)) {
            for (var conn : WireGraph.get(fromV)) {
                if (Cliques.get(to).contains(conn)) {
                    cutCnt++;
                }
            }
        }
        return cutCnt;
    }

    int FindFather(Map<Integer, Integer> Hierachy, int name) {
        if (Hierachy.get(name).equals(name)) {
            return name;
        }
        return Hierachy.compute(name, (k, e) -> FindFather(Hierachy, Hierachy.get(e)));
    }

    Map<Integer, Set<Integer>> Karger(long seed, long timeout) {
        var now = System.currentTimeMillis();
        Random rand = new Random(seed);
        Map<Integer, Set<Integer>> Cliques = new HashMap<>();
        Map<Integer, Integer> Hierachy = new HashMap<>();
        Set<Integer> NodesHavingEdge = new HashSet<>(WireGraph.keySet());
        HashMap<Integer, Set<Integer>> visited = new HashMap<>();
        WireGraph.keySet().forEach(e -> {
            visited.put(e, new HashSet<>());
            Cliques.put(e, new HashSet<>(Set.of(e)));
            Hierachy.put(e, e);
        });
        while (Cliques.size() != 2) {
            if (System.currentTimeMillis() - now > timeout) {
                break;
            }
            // Get father
            List<Integer> NodeList = NodesHavingEdge.stream().toList();
            Integer father = NodeList.get(rand.nextInt(NodeList.size()));
            // Get child
            List<Integer> ChildList = WireGraph.get(father).stream()
                    .filter(child -> !visited.get(father).contains(child) && !visited.get(child).contains(father))
                    .toList();
            if (ChildList.isEmpty()) {
                NodesHavingEdge.remove(father);
                continue;
            }
            Integer child = ChildList.get(rand.nextInt(ChildList.size()));
            visited.get(father).add(child);
            visited.get(child).add(father);
            int FatherCliqueName = FindFather(Hierachy, father), ChildCliqueName = FindFather(Hierachy, child);
            Hierachy.put(ChildCliqueName, FatherCliqueName);
            if (FatherCliqueName != ChildCliqueName) {
                Cliques.get(FatherCliqueName).addAll(Cliques.get(ChildCliqueName));
                Cliques.remove(ChildCliqueName);
            }
        }
        return Cliques;
    }

    public void Solution1(BufferedReader input) throws Exception {
        readFile(input);
        // long timeout = 70;
        // while (true) {
        // long seed = System.currentTimeMillis();
        // var res = Karger(seed, timeout);
        // if (CutCount(res) == 3) {
        // System.out.println("Using seed: " + seed);
        // System.out.println("Solution 1: " + res.values().stream().mapToInt(s -> s.size()).reduce(1, (i, v) -> i * v));
        // break;
        // }
        // }
        // 1747570281531 - 40ms
        System.out.println("Solution 1: " + Karger(1747570281531L, Long.MAX_VALUE).values().stream().mapToInt(s -> s.size()).reduce(1, (i, v) -> i * v));
    }

    public void Solution2(BufferedReader input) throws Exception {
        System.out.println("\t\t\t------All 50 stars AoC Finished!!!!------");
    }

    public static void main(String[] args) throws Exception {
        MinimumCut Day25 = new MinimumCut();
        Day25.Solution1(new BufferedReader(new FileReader("Day25/input.txt")));
        Day25.Solution2(new BufferedReader(new FileReader("Day25/input.txt")));
    }
}

// var sorted =
// crossCnt.entrySet().stream().sorted(Comparator.comparing(Entry<String,
// Integer>::getValue).reversed());
// var itr = sorted.iterator();
// System.out.println(itr.next().getKey() + "," + itr.next().getValue());
// System.out.println(itr.next().getKey() + "," + itr.next().getValue());
// System.out.println(itr.next().getKey() + "," + itr.next().getValue());
// System.out.println(itr.next().getKey() + "," + itr.next().getValue());
// System.out.println(itr.next().getKey() + "," + itr.next().getValue());
// System.out.println(itr.next().getKey() + "," + itr.next().getValue());
// BufferedWriter output = new BufferedWriter(new FileWriter("output.txt"));
// for (Entry<String, Integer> e :
// crossCnt.entrySet().stream().sorted(Comparator.comparing(Entry<String,
// Integer>::getValue).reversed()).toList()) {
// output.write(e.getKey() + "," + e.getValue());
// output.newLine();
// }
// output.close();// The buffer is never flushed before closed.

File: Java/2023/main/year2023/Day25/MinimumCut.kt
package year2023.Day25

import JavaDataModel.AoCSolution
import JavaDataModel.ISolution
import java.io.*
import java.util.*
import JavaDataModel.inputFileName

@AoCSolution(day = 25)
class MinimumCut : ISolution {
    fun readFile(input: BufferedReader) {
        input.use { data ->
            data.lines().forEach { l ->
                val split = l.split(":")
                val inWire = Hasher(split[0].trim())
                split[1].trim().split(" ").asSequence().map(this::Hasher).forEach { wHash ->
                    WireGraph.getOrPut(inWire) { HashSet() }.add(wHash)
                    WireGraph.getOrPut(wHash) { HashSet() }.add(inWire)
                }
            }
        }
    }

    fun Hasher(WireName: String) = WireName.fold(0) { i, c -> i * 26 + c.code - 'a'.code }

    val WireGraph: MutableMap<Int, MutableSet<Int>> = HashMap()

    // fun CutCount(Cliques: Map<Int, Set<Int>>) = when {
    //     Cliques.size != 2 -> -1
    //     else -> {
    //         val (from, to) = Cliques.values.take(2)
    //         from.sumOf { WireGraph[it]?.count { conn -> conn in to } ?: 0 }
    //     }
    // }

    fun FindCliqueRoot(CliqueRootMap: HashMap<Int, Int>, name: Int): Int =
        CliqueRootMap[name]!!.let { f ->
            when {
                CliqueRootMap[f] == name -> f
                else -> CliqueRootMap.compute(name) { k, v -> FindCliqueRoot(CliqueRootMap, f) }!!
            }
        }

    fun Karger(seed: Long, timeout: Long): Map<Int, Set<Int>> {
        val now = System.currentTimeMillis()
        val rand = Random(seed)
        val Cliques: MutableMap<Int, MutableSet<Int>> = HashMap()
        val CliqueRootMap: HashMap<Int, Int> = HashMap()
        val NodesHavingEdge: MutableSet<Int> = HashSet(WireGraph.keys)
        val visited: MutableMap<Int, MutableSet<Int>> = HashMap()
        WireGraph.keys.forEach { e ->
            visited.put(e, HashSet())
            Cliques.put(e, mutableSetOf(e))
            CliqueRootMap.put(e, e)
        }
        while (Cliques.size != 2) {
            if (System.currentTimeMillis() - now > timeout) {
                break
            }
            // Get father
            val NodeList = NodesHavingEdge.toList()
            val root = NodeList[rand.nextInt(NodeList.size)]
            // Get child
            val ChildList = WireGraph[root]!!.filter { child ->
                !visited[root]!!.contains(child) && !visited[child]!!.contains(root)
            }.toList()
            if (ChildList.isEmpty()) {
                NodesHavingEdge.remove(root)
                continue
            }
            val child = ChildList[rand.nextInt(ChildList.size)]
            visited[root]!!.add(child)
            visited[child]!!.add(root)
            val CliqueRootId = FindCliqueRoot(CliqueRootMap, root)
            val CliqueChildId = FindCliqueRoot(CliqueRootMap, child)
            CliqueRootMap.put(CliqueChildId, CliqueRootId)
            if (CliqueRootId != CliqueChildId) {
                Cliques[CliqueRootId]!!.addAll(Cliques[CliqueChildId]!!)
                Cliques.remove(CliqueChildId)
            }
        }
        return Cliques
    }

    override fun Solution1(input: BufferedReader) {
        readFile(input)
        //val timeout = 70L
        //while (true) {
        //    val seed = System.currentTimeMillis()
        //    val res = Karger(seed, timeout)
        //    if (CutCount(res) == 3) {
        //        println("Using seed: ${seed}, time cost ${System.currentTimeMillis() - seed}")
        //        println("Solution 1: " + res.values.map { it.size }.fold(1) { i, v -> i * v })
        //        return
        //    }
        //}
        // Kotlin 1759684678660 - 20 ms
        // Java - 1747570281531 - 40 ms
        // val res = Karger(1747570281531L, Long.MAX_VALUE)
        // println(CutCount(res)) // This is 4 in Kotlin, but 3 in Java
        // Kotlin seems to have modified either HashMethod or Random generator
        println("Solution 1: " + Karger(1759684678660L, Long.MAX_VALUE).values.asSequence().map { s -> s.size }
            .fold(1) { i, v -> i * v })
    }

    override fun Solution2(input: BufferedReader) {
        println("\t\t\t------All 50 stars AoC Finished!!!!------")
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            MinimumCut().let { day25 ->
                BufferedReader(FileReader(day25.inputFileName!!)).use { day25.Solution1(it) }
                BufferedReader(FileReader(day25.inputFileName!!)).use { day25.Solution2(it) }
            }
        }
    }
}
